test/widgets/chat/drive_chat_page_test.dart

```diff
@@ -0,0 +1,333 @@
+import '../../util/factories/drive_factory.dart';
+import '../../util/factories/message_factory.dart';
+import '../../util/factories/model_factory.dart';
+import '../../util/factories/profile_factory.dart';
+import '../../util/factories/ride_factory.dart';
+import '../../util/mock_server.dart';
+import '../../util/pump_material.dart';
+import '../../util/request_processor.dart';
+import '../../util/request_processor.mocks.dart';
+
+void main() {
+  late Drive drive;
+  final MockRequestProcessor processor = MockRequestProcessor();
+  final Profile profile = ProfileFactory().generateFake(id: 1);
+
+  setUpAll(() {
+    MockServer.setProcessor(processor);
+  });
+
+  group('Has Stream Subscription', () {
+    setUp(() async {
+      whenRequest(processor).thenReturnJson([]);
+    });
+    testWidgets('when Drive has rides with active chat', (WidgetTester tester) async {
+      drive = DriveFactory().generateFake(
+        driverId: profile.id,
+        rides: [RideFactory().generateFake(chatId: 1, status: RideStatus.approved)],
+      );
+      await pumpMaterial(tester, DriveChatPage(drive: drive));
+      await tester.pump();
+      verifyRequest(
+        processor,
+        urlMatcher: equals('/rest/v1/messages?select=%2A&order=created_at.desc.nullslast'),
+      ).called(1);
+      final List<RealtimeChannel> subscription = SupabaseManager.supabaseClient.getChannels();
+      expect(subscription.length, 1);
+      expect(subscription[0].topic, 'realtime:public:messages:1');
+    });
+
+    testWidgets('not when Drive has no rides', (WidgetTester tester) async {
+      drive = DriveFactory().generateFake(
+        driverId: profile.id,
+        rides: [],
+      );
+      await pumpMaterial(tester, DriveChatPage(drive: drive));
+      await tester.pump();
+      verifyRequestNever(
+        processor,
+        urlMatcher: equals('/rest/v1/messages?select=%2A&order=created_at.desc.nullslast'),
+      );
+      final List<RealtimeChannel> subscription = SupabaseManager.supabaseClient.getChannels();
+      expect(subscription.length, 0);
+    });
+
+    testWidgets('not when Drive has rides without active chat', (WidgetTester tester) async {
+      drive = DriveFactory().generateFake(
+        driverId: profile.id,
+        rides: [RideFactory().generateFake(status: RideStatus.pending)],
+      );
+      await pumpMaterial(tester, DriveChatPage(drive: drive));
+      await tester.pump();
+      verifyRequestNever(
+        processor,
+        urlMatcher: equals('/rest/v1/messages?select=%2A&order=created_at.desc.nullslast'),
+      );
+      final List<RealtimeChannel> subscription = SupabaseManager.supabaseClient.getChannels();
+      expect(subscription.length, 0);
+    });
+  });
+
+  testWidgets('Shows chats from Rides that have a active Chat', (WidgetTester tester) async {
+    drive = DriveFactory().generateFake(
+      driverId: profile.id,
+      rides: [
+        RideFactory().generateFake(chatId: 1, status: RideStatus.approved),
```

Comment:
why do these chatIds need to be hard-coded? You are not even using them later. But generally, I think we should hard-code way less of these ideas anyways.

Reply:
because otherwise i could have the same ids then the test would fail.

Reply:
The factory generates ids as a random number between 1 and 10000000. I don't think that's likely :D

Reply:
Yes but still possible and then you have a test that fails and don't know why

Reply:
You would do the test locally and see that it was just fluky... also these odds are insanely low if you calculate it out

Reply:
and we can easily add one more zero if the number is not high enough lol

Reply:
but then the possibility would still be there, i just don't see why that's nessecary.

Reply:
I have to take Fynn's side here. Tests are never going to be 100.0% accurate, and I think going for the last literal 0.0000001% is the wrong incentive here (for comparison, the most advanced autonomous vehicles have an injury collision avoidance performance of 99.9999%, which would make them less safe than this test). That comes at the cost of a little wasted space where the reader has to spend some time thinking about why the IDs were specified, which is not the point of the test. It makes the test less focused, less readable and less clean

Reply:
okay I'm convinced by the argument with the autonomous vehicle.